components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/artifact/SAMLSSOArtifactResolutionService.java

```diff
@@ -0,0 +1,196 @@
+        this.authenticatorProperties = authenticatorProperties;
+        this.tenantDomain = tenantDomain;
+    }
+
+    /**
+     * Create a SAML artifactResolveObject based on given samlArt parameter and wrapped into a soapRequest
+     * Send the soapRequest to the Artifact Resolve Endpoint
+     *
+     * @param samlArt SAML Artifact reference needed to get the actual data
+     * @return ArtifactResponse
+     */
+    public ArtifactResponse getSAMLArtifactResponse(String samlArt) throws ArtifactResolutionException {
+
+        validateArtifactResolveConfig();
+        ArtifactResolve artifactResolve = generateArtifactResolveReq(samlArt);
+        return sendArtifactResolveRequest(artifactResolve);
+    }
+
+    /**
+     * Create SAML ArtifactResolve Object and sign
+     *
+     * @param sReceivedArtifact object reference to actual data
+     * @return SAML ArtifactResolve Object
+     */
+    public ArtifactResolve generateArtifactResolveReq(String sReceivedArtifact) throws ArtifactResolutionException {
+
+        ArtifactResolve artifactResolve = createArtifactResolveObject(sReceivedArtifact);
+        if (SSOUtils.isArtifactResolveReqSigned(authenticatorProperties)) {
+            if(log.isDebugEnabled()) {
+                log.debug("Signing artifact resolve request.");
+            }
+            signArtifactResolveReq(artifactResolve);
+        }
+        return artifactResolve;
+    }
+
+    /**
+     * Send Artifact Resolve Request to Artifact Resolution Service
+     *
+     * @param artifactResolve Artifact Resolve Request
+     * @return ArtifactResponse
+     * @throws ArtifactResolutionException
+     */
+    public ArtifactResponse sendArtifactResolveRequest(ArtifactResolve artifactResolve) throws ArtifactResolutionException {
+
+        SAMLSSOSoapMessageService soapMessageService = new SAMLSSOSoapMessageService();
+        Envelope envelope = soapMessageService.buildSOAPMessage(artifactResolve);
+        String envelopeElement;
+        try {
+            envelopeElement = SSOUtils.marshall(envelope);
+        } catch (SAMLSSOException e) {
+            throw new ArtifactResolutionException("Encountered error marshalling message into its DOM representation", e);
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug("Artifact Resolve Request as a SOAP Message: " + envelopeElement);
+        }
+
+        String artifactResponseString = soapMessageService.sendSOAP(envelopeElement, SSOUtils.getArtifactResolveUrl
+                (authenticatorProperties));
+        Pattern p = Pattern.compile("<samlp:ArtifactResponse.+</samlp:ArtifactResponse>", Pattern.DOTALL);
+        Matcher m = p.matcher(artifactResponseString);
+        if (m.find()) {
+            if (log.isDebugEnabled()) {
+                log.debug("Artifact Response: " + m.group(0));
+            }
+            try {
+                ArtifactResponse artifactResponse = (ArtifactResponse) SSOUtils.unmarshall(m.group(0));
```

Comment:
Get the response object assigned to the variable and use that in the debug log and to unmarshall.
Also, in which format does the unmarshalled response gets printed when you log it in debug.